connector/github/github.go

```diff
@@ -411,12 +416,89 @@ func (c *githubConnector) user(ctx context.Context, client *http.Client) (user,
 		return u, fmt.Errorf("%s: %s", resp.Status, body)
 	}
 
-	if err := json.NewDecoder(resp.Body).Decode(&u); err != nil {
+	if err = json.NewDecoder(resp.Body).Decode(&u); err != nil {
 		return u, fmt.Errorf("failed to decode response: %v", err)
 	}
+
+	if u.Email == "" {
+		if u.Email, err = c.userEmail(ctx, client); err != nil {
+			return u, err
+		}
+	}
 	return u, nil
 }
 
+type userEmail struct {
+	Email      string `json:"email"`
+	Verified   bool   `json:"verified"`
+	Primary    bool   `json:"primary"`
+	Visibility string `json:"visibility"`
+}
+
+// userEmail queries the GitHub API for a users' email information using the
+// provided client. Only returns the users' verified, primary email (private or
+// public).
+//
+// The HTTP client is expected to be constructed by the golang.org/x/oauth2 package,
+// which inserts a bearer token as part of the request.
+func (c *githubConnector) userEmail(ctx context.Context, client *http.Client) (string, error) {
+
+	// https://developer.github.com/v3/#pagination
+	reNext := regexp.MustCompile("<(.*)>; rel=\"next\"")
+	reLast := regexp.MustCompile("<(.*)>; rel=\"last\"")
+	// https://developer.github.com/v3/users/emails/#list-email-addresses-for-a-user
+	apiURL := c.apiURL + "/user/emails"
+
+	var emails []userEmail
+	for {
+		req, err := http.NewRequest("GET", apiURL, nil)
+		if err != nil {
+			return "", fmt.Errorf("github: new req: %v", err)
+		}
+		req = req.WithContext(ctx)
+		resp, err := client.Do(req)
+		if err != nil {
+			return "", fmt.Errorf("github: get URL %v", err)
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return "", fmt.Errorf("github: read body: %v", err)
+			}
+			return "", fmt.Errorf("%s: %s", resp.Status, body)
+		}
+
+		if err := json.NewDecoder(resp.Body).Decode(&emails); err != nil {
+			return "", fmt.Errorf("failed to decode response: %v", err)
+		}
+
+		for _, email := range emails {
```

Comment:
What's the ordering here? If the user has multiple emails how do we determine which one we choose? Does GitHub rank these in some way?

If GitHub's doesn't guarantee an ordering we should at least short them. But I wonder if there's a better way.

cc @simonbyrne